Tenants, External Users, and tenant based JWT with Audit Events

johnswanson opened the pull request:
This is the rebased version of tenants. Pushing to a new branch to avoid clobbering the existing one.

johnswanson committed:
Tenants, External Users, and tenant based JWT with Audit Events

Allow the creation of tenant users

While working on this, I realized that my approach to the "All External
Users" group and renaming the "All Users" group was untenable.

I had been planning to just update these things in the database (create
AEU, rename AU to AIU) in the setter function for the `:use-tenants`
setting. The issue here is caching: we don't want to need to actually
reach out to the database every time we need to fetch the All Users or
All External Users, but it would also be quite unfortunate if one
Metabase instance in a cluster had an outdated view of, e.g. the ID of
"All External Users".

(Although - I'm not really sure how much we use the *name* of "All
Users" when we fetch `(all-users)`, so that might be worth a look.)

For now, I've modified things to just update things dynamically in the
`GET /api/permissions/group` and `GET /api/permissions/group/:id`
endpoints: we show/hide the "All External Users" group, and rename the
"All Users" group, based on whether Tenants are enabled.

Allow the creation of tenants

Listing tenants / tenant members will be a followup PR

lint

doh missing require

fix tests

add a test for enabling tenants changing groups

fix tests, slight change to tenant creation API

kick ci

fix another test

give minimum permissions to ext groups

fix more tests

impl user stategy setting ui

more API endpoints

add more tests, fix a very silly migration

slug validation

fix mariadb/mysql

lint migrations file

kondo

fix test

replace tenant members API with `/user?tenant_id=`

also realized that the `/user/recipients` endpoint needed to be modified
for tenants: tenant users should not see recipients that are not part of
the same tenant. (reminder for later: we will need to add safeguards
around actually creating those pulses, not just assume the API won't
send bad recipients!!)

IF you are an admin, `/api/user` respects a `tenant_id` query param and returns:

- "normal" (non-tenant) users if it's `nil`

- specific-tenant users if not

If you are *not* an admin, `/api/user` ignores the `tenant_id` passed
and only returns users in the same tenant.

`/api/user/recipients` will:

- return ALL users if you are a superuser (both tenant and non-tenant)

- return only users in the SAME tenant as you if not

doh

tenant crud wip

integrate endpoints, fix type errors

Allow getting all external users

previously you could do only

```
GET /api/user                => internal users
GET /api/user?tenant_id=null => internal users
GET /api/user?tenant_id=123  => users in tenant 123
```

now you can do

```
GET /api/user                  => internal users
GET /api/user?tenant_id=null   => internal users
GET /api/user?tenant_id=123    => users in tenant 123
GET /api/user?tenancy=all      => internal + external users
GET /api/user?tenancy=internal => internal users
GET /api/user?tenancy=external => external users (in all tenants)
```

If you specify both `tenancy` and `tenant_id` (regardless of the values
passed) you'll get a 400.

fix my bad rebase

fix bad rebase more

remove #p

mark tenants as `is_active` or not

initial external users management

remove permissions check on the trash endpoint

I don't know why this was added in the first place (I mean, I think I
added it, but I don't know why I did).

The trash is always accessible to everyone. We gate the visibility of
things *in* the trash, but the trash itself is always readable.

no personal collection for tenant users

remove tenant query inside oss

tenant user routes

make it so that you cannot remove users from the external users group

do not send email to tenant users

no login for inactive tenants

disable collection permissions, don't show sent email modal on user success with SMTP

remove all users group has higher permissions warnings for external users group

moving is external checks behind plugins

unit test cleanup, add simple test

read only application permissions for external users

set tenant on login with JWT

incredibly hacky tenant attribute

separate `:attributes` from `:login_attributes`

thinking here is that `login_attributes` are the things you set on the
user, where `attributes` are all the attributes of the user including
things like tenant slug or maybe someday locale etc.

fix build

fix token check

create a tenant collection in a migration

this works, but the tenant collection is always visible to admins, even
if the setting is off.

hide tenant collection from normal users

fixing build and FE rebase

Delete stray "howdy"

Lol, sorry about that.

fix tests

throw instead of silent failure on no tenant

superuser can't be tenant

fix up tenant collections

tenant collection now includes subcollections.

Tenant collections are hidden if the feature isn't enabled.

tenant collections marked in API

bugfix and current user can write any coll

don't add `attributes` key to all users

just the current user on login

invalidate groups when changing user strategy

read only user

{{ tenant_slug }} user attribute

fix be test

rip out tenant collection stuff on BE

fix BE tests

lint

fix magic string

no tenantifying existing users for now

also clean up a complicated, confusing fn.

starting e2e

fixing e2e, adjusting @tenant user attribute

fix permissions editor, add tenant.name e2e test

allow more than one user per tenant

:facepalm:

allow getting only active tenants

*actually* don't replace existing values with nil

getting e2e ready for tenant user provisioning

tenants management ux

e2e coverage on sandboxing, deactivating tenants

replace {{tenant_slug}} with `@tenant.slug`

filter out other-tenant users from pulse creation

rename any existing `@.+` user attributes

prohibit new user attributes starting with `@`

fix migration

ugh, forgot to actually commit this file

add more tests

kondo

use the `tenants` feature

fix a test - missed one `:advanced-permissions`

e2e test coverage, updating tenant attribute to @tenant.slug

Partial Type Fixes

types green

unit tests

adding groups to e2e test, maybe fixing tenant feature check?

Add tenants to tracked snowplow features

Fixing Admin ui

Modernizing tenants page

separate external group in permissions sidebar

permissions editor updates

hardcode that *everyone* has trash access

shame... bad imports

adjusting Main Navbar. using globe icon

CR response

fix tests

fix attributes when tenants are disabled

lock up the 'all external users' collection

tenants route guard

add tests for missing feature and oss

fixing some unit tests, and type errors

set token features in default.cy.snap

unit and e2e test adjustment

another pass on e2e

fix e2e and build

fix a bad merge

Green CI?

reset token after restoring DB

move a test to fix the test

kondo

Move `/api/ee/tenants` to `/api/ee/tenant`

PR Feedback, part 1

move route functions into ee

ADM-1105 permissions editor adjustments for external users

ADM 1104 do not show invite button on external users when there is no tenant

adjusting unit tests

fix no results message for deactivated users page

center no results message

Fix migrations IDs / related tests

adjusting no results message for tehants

fixing unit tests, adding e2e test for permissions

chore: add timestamp columns to tenants (#60575)

* chore: add timestamp columns to tenants

* fixup! add hooks

* kick ci

Auto provision tenants, respect tenants feature (#60571)

deactivate tenant users along with their tenants (#60552)

* deactivate tenant users along with their tenants

Tenant users from deactivated tenants were already prevented from
logging in (we checked the activation status of their tenant) but for
bookkeeping it's probably best to actually deactivate the users
themselves along with their tenants, and reactivate them if the tenant
is reactivated.

* invalidate users when updating a tenant, adjust external users list screen

* e2e test

---------

Co-authored-by: Nick Fitzpatrick <nickfitz.582@gmail.com>

audit events for tenants (#60647)

johnswanson committed:
Revert "Delete unused string util from metabase lib (#64389)"

This reverts commit 0af38ef8f021726985eacccddffca8f0b43233f8.

johnswanson committed:
fix the build

johnswanson committed:
fix tests

heypoom committed:
fix rebase error in invite button

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix more tests

johnswanson committed:
fix another test

johnswanson committed:
Fix another couple tests

johnswanson committed:
and more

johnswanson committed:
and more

heypoom committed:
Close the modal to create shared tenant collection after it is created (#66049)

johnswanson committed:
Fix more tests, nearly all e2e tests passing

johnswanson committed:
a couple more tests

johnswanson committed:
couple more tests

johnswanson committed:
last (???) few tests

heypoom committed:
Remove edit permissions in ellipsis menu when viewing shared tenant collection (#66042)

heypoom committed:
Suggest external users only when adding tenant group members (#66046)

johnswanson committed:
Dedicated Tenant Collections

edpaget committed:
feat: separate namespace for tenant collections (#66102)

edpaget committed:
fixup! FE build

johnswanson committed:
More fixes

johnswanson committed:
Tests and more fixes

edpaget committed:
feat: add remote-sync support for tenant collections (#66272)

* feat: add remote-sync support for tenant collections

* Update permissions.clj

* fixup! tests

* fixup! enterprise tests

* fixup! mysql/h2

edpaget committed:
EMB-1011: add default read perms for namespace

johnswanson committed:
Fix rebase error

johnswanson committed:
Fix tests (#66403)

* Fix tests

Not exactly sure how these got merged, I think maybe they were hidden by
GH issues.

* Fix kondo

* fix apparent rebase error

edpaget committed:
fixup! EMB-1065 fix namespace filtering

edpaget committed:
fixup! some bad perms tests

edpaget committed:
fixup! remote sync dep checking across collections

edpaget committed:
fixup! remove blank test

edpaget committed:
fixup! missing tenant setting

edpaget committed:
fixup! add premium feature declaration

edpaget committed:
fixup! get test working with h2

edpaget committed:
fixup! revert h2 fix

edpaget committed:
fixup! revert h2 fix

edpaget committed:
fixup! real h2 fix

johnswanson committed:
rename

johnswanson committed:
Fix query for shared tenant collections

johnswanson committed:
Give tenant users a personal collection

edpaget committed:
fixup! UXW-2386 don't pass `tables` model into tenant root

edpaget committed:
fixup! UXW-2436 build correct path for namespaced collections

edpaget committed:
Remove test of non-existent fn

heypoom committed:
Fix failing frontend unit and e2e tests in tenants feature branch (#66243)

Co-authored-by: John Swanson <john.d.swanson@gmail.com>

heypoom committed:
fix failing frontend ESLint errors

heypoom committed:
Prefix users and groups as "Internal" in admin settings when tenants are active (#66169)

heypoom committed:
Update design of the user strategy modal for tenants (#66182)

Co-authored-by: Mahatthana (Kelvin) Nomsawadi <me@bboykelvin.dev>

heypoom committed:
Add multi-tenancy setup guide to embedding hub (#66113)

heypoom committed:
Update design of the tenants listing and internal user page (#66193)

johnswanson committed:
Fix BE tests

johnswanson committed:
another be test

johnswanson committed:
kondo/fmt

heypoom committed:
Add specific error messages for JWT tenant claim errors (#66172)

johnswanson committed:
fix running jwt tests standalone

edpaget committed:
feat: per collection setting (#66475)

* feat: per collection setting

* feat: remove unsynced collections

* fixup! tests

* fixup! lints

* fixup! test

edpaget committed:
Remove duplicated tenant permissions

johnswanson committed:
handle disabling feature more cleanly

when tenants is turned off:

- disable all tenant users

- reject any logins from tenant users

- reject any JWTs that specify a tenant

edpaget committed:
Add tenant entity picker e2e tests. Fix one bug surfaced

iethree commented on frontend/src/metabase/common/components/Pickers/CollectionPicker/hooks.ts:
```diff
@@ -139,6 +141,33 @@ export const useRootCollectionPickerItems = (
+        namespace: "shared-tenant-collection",
+        here: ["collection", "card", "dashboard"],
+        description: null,
+        can_write: true,
```
this should prevent saving in this collection, without preventing navigating into it

```suggestion
        can_write: false,
```

heypoom committed:
uxw-2456: remove drop shadows from dedicated tenant collection cards

heypoom committed:
Hide tenant collections permissions tab when tenants are disabled (#66537)

heypoom committed:
Hide create shared tenant collection button on sidebar for non-admin users (#66538)

heypoom committed:
Hide tenant collections sidebar section when no collections exist for non-admins (#66539)

heypoom committed:
Fix dedicated tenant collection root in collection picker as tenant user (#66549)

edpaget committed:
EMB-2459: Fix restrict where collections can be moved to

heypoom committed:
fix frontend type-check: only set namespace if it is defined

johnswanson committed:
EMB-1093: Fix not passing a tenant attribute

Fixes [EMB-1093: If you enable multi-tenants and don't add JWT
attribute, it will
break](https://linear.app/metabase/issue/EMB-1093/if-you-enable-multi-tenants-and-dont-add-jwt-attribute-it-will-break)

edpaget committed:
Fixup working e2e entity picker tests

edpaget committed:
fixup! mysql namespace tests

edpaget committed:
fixup! tenant group mapping tests

edpaget committed:
fixup! ws

johnswanson committed:
tree endpoint allows specifying multiple nses

johnswanson committed:
fix broken test

heypoom committed:
Show dedicated tenant collections in collection pickers for admins (#66583)

heypoom committed:
feat(sdk-bundle): add tenants ee plugin (#66586)

edpaget committed:
fixup! bad ns clause audit feature is off

edpaget committed:
fixup! UXW-2473 support non-string attributes

heypoom committed:
Hide tenant collections section when there are no visible shared collections (#66597)

edpaget committed:
Add tenant cards to search

edpaget committed:
fixup! tenant collection cleanup

npretto committed:
EMB-1082 rename strings in the ui (#66603)

npretto committed:
make tenant field required (#66516)

johnswanson committed:
change `tenant` key to `@tenant`

- if the feature is off, you can still log in if you have the `@tenant`
key set. It will be silently ignored, but a warning will be logged.

- if the feature is on, `@tenant` is a drop-in replacement for `tenant`.

- we'd previously prevented setting user attributes that start with `@`.
This updates `remove-invalid-attributes` to do the same with JWT/SAML
attributes.

johnswanson committed:
add SAML tests

johnswanson committed:
Expose SAML attribute for tenant

johnswanson committed:
External users can't use MB if feature disabled

Plus add tests for:

- tenant user, feature disabled

- tenant user, feature unavailable

johnswanson committed:
Add missing test files

heypoom committed:
Multi-stage embedding data picker should query models and questions from tenant collections (#66598)

heypoom committed:
Fix external group tooltip referencing All Internal Users (#66654)
